titus-server-master/src/main/java/com/netflix/titus/master/jobmanager/service/service/ScaleDownEvaluator.java

```diff
@@ -177,6 +183,36 @@ private static List<ServiceJobTask> selectTasksToTerminateInEquivalenceGroup(Lis
         return tasksToKill;
     }
 
+    private static List<ServiceJobTask> selectTasksToTerminateInEquivalenceGroupFavorBinPacking(List<ServiceJobTask> allTasks, int targetTerminateCount, TitusRuntime titusRuntime) {
+        List<ServiceJobTask> tasksToKill = new ArrayList<>();
+        Region region = new Region(allTasks);
+
+        while (tasksToKill.size() < targetTerminateCount && region.hasMoreTasks()) {
+            // Step 1: select smallest groups of tasks to scale down
+            // smallest number of tasks on an agent in any zone
+            int smallestGroup = region.getSmallestTaskGroup();
+
+            // Step 2: kill tasks from the smallest groups, trying to maintain zone balancing
+            // There is increased likelihood that this way we will give agents with a smaller number of tasks
+            // a chance of being scaled-down.
+            boolean hasMore = true;
+            while (tasksToKill.size() < targetTerminateCount && hasMore) {
+                Optional<Zone> zoneOpt = region.getSmallestZoneWithTaskGroupSize(smallestGroup);
```

Comment:
How does this maintain zone balancing? 
If we really want to maintain zone balancing (more than binbacking) then I think this should reuse `getLargestZoneWithTaskGroupSize`.

Reply:
I think the name of this function needs to be improved. 
Given two zones with multiple agents where both the zones have an agent with the smallest number of tasks/agent, we'd prefer to terminate tasks on the agent in a zone that has more overall tasks so we end up reducing the disparity between zones.

Reply:
I don't think zone balance should be a responsibility or consideration of this component at all.  Zone balance seems like a responsibility of something in fleet management land.  If, due to binpacking, we are always removing machines from a particular zone, generating zone imbalance, then fleet management should prefer to add machines in the same zone.

Reply:
Think about it this way.  We can't address every optimization at every layer so we have to make clear decisions about component responsibility.  Why are we tackling zone balancing here?  Why not node health (prefer descheduling on nodes which are frequently unhealthy)?  Why not resource utilization (prefer descheduling on nodes which are excessively oversubscribed)?  Why not deployment plans (prefer descheduling on nodes part of an immutable deploy)?

All these optimizations could help bin-packing or any of a variety of fleetwide goals.  But we can't responsibly embed these heuristics into a single component, and then try to embed them again in a different component.  Scheduling and descheduling sides need to have a shared view of what nodes are to be avoided/preferred.

A single place needs to synthesize all our fleet management policy goals so that scheduling and descheduling sides can cooperate in a debuggable and maintainable fashion.